design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,278 @@
+
+This checklist contains actions which must be completed before a PR implementing this design can be merged.
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
+- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]
+
+## Summary
+
+cert-manager has the ability to set the owner reference field in generated Secret resources. The option is global, and takes the form of the flag `--enable-certificate-owner-ref` set in the cert-manager controller Deployment resource.
+
+Let us take an example of Certificate resource:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: cert-1
+  namespace: ns-1
+  uid: 1e0adf8
+spec:
+  secretName: cert-1
+```
+
+When `--enable-certificate-owner-ref` is passed to the cert-manager controller, when issuing the X.509 certificate, cert-manager will create a Secret resource that looks like this:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cert-1
+  namespace: ns-1
+  ownerReferences:
+    - controller: true
+      blockOwnerDeletion: false
+      uid: 1e0adf8
+      name: cert-1
+      kind: Certificate
+      apiVersion: cert-manager.io/v1
+data:
+  tls.crt: "..."
+  tls.key: "..."
+  ca.crt: "..."
+```
+
+The proposition is to add a new field `certificateOwnerRef` to the Certificate resource:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+spec:
+  secretName: cert-1
+  cleanupPolicy: [Delete|Never] # ✨
+```
+
+The new field `cleanupPolicy` has three possible values:
+
+1. When "empty", the behavior will default to not creating an owner reference on the Secret resource, unless `--enable-certificate-owner-ref` is passed.
+2. When `Delete`, the default behavior as described in the "empty" case is overridden and the owner reference is always created on the Secret resource.
+3. When `Never`, the default behavior as described in the "empty" case is overridden and the owner reference is never created on the Secret resource.
+
+> At first, the proposed field was named `certificateOwnerRef` and was a
+> nullable boolean. James Munnelly reminded us that the Kubernetes API
+> never uses boolean fields, and instead uses the string type with
+> "meaningful values". On top of being more readable, it also makes the
+> field extensible.
+
+## Use-cases
+
+Flant manages certificates for users, and has hit a Kubernetes apiserver limitation where too many left-over Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.
```

Comment:
```suggestion
Flant manages certificates for users, and has hit a Kubernetes apiserver limitation where too many leftover Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.
```